src/callbacks/perturb_weights.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/comm_impl.hpp"
+#include "lbann/callbacks/perturb_weights.hpp"
+#include "lbann/proto/proto_common.hpp"
+#include "lbann/utils/serialize.hpp"
+#include "lbann/weights/data_type_weights.hpp"
+
+#include <callbacks.pb.h>
+
+
+namespace lbann {
+namespace callback {
+
+perturb_weights::perturb_weights(std::string output_name,
+                           El::Int batch_interval)
+  : callback_base(batch_interval),
+    m_output_name(std::move(output_name)) {
+    m_output = nullptr;
+}
+
+perturb_weights::perturb_weights()
+  : perturb_weights("",0)
+{}
+
+template <class Archive>
+void perturb_weights::serialize(Archive & ar) {
+  ar(::cereal::make_nvp(
+       "BaseCallback",
+       ::cereal::base_class<callback_base>(this)),
+     CEREAL_NVP(m_output_name));
+}
+
+void perturb_weights::setup(model* m) {
+   for (auto* w : m->get_weights()) {
+      if(w->get_name() == m_output_name){
+        m_output = w;
+        break;
+      }
+   }
+    if (m_output == nullptr) {
+      LBANN_ERROR("Current implementation of callback \"", name(), "\" "
+                  "requires a weight object to perturb");
+    }
+}
+
+void perturb_weights::on_batch_begin(model* m) {
+  const auto& c = m->get_execution_context();
+  if (m_output != nullptr && 
+      c.get_step() % m_batch_interval == 0 &&
+      c.get_step() >= 0) {
+    perturbed(*m);
+  }
+}
+
+void perturb_weights::perturbed(model& m){
+
+  auto* comm = m.get_comm();
+
+  // Useful constants
+  constexpr DataType zero = 0;
+  constexpr DataType one = 1;
+  const auto range = 3;
+  const auto thres = 0.5;
+  const auto lower = 0.3;
+  const auto upper = 0.7;
+  
+  // RNG
+  auto& gen = get_generator();
+  std::normal_distribution<DataType> dist(zero, one);
+
+
+  for (auto* w : m.get_weights()) {

```

Comment:
I don't think we need to iterate through the weights since we've already found `m_output` in `setup`. Alternatively, we could remove `m_output` as a member variable and search for the weights every time this callback is activated. I prefer this approach since it makes the callback stateless.

Reply:
Updated.